### PR TITLE
UI console: Restore tty settings, do not force ECHO after prompt

### LINKED ISSUE #2373: [rt.openssl.org #3935] TTY echo flag not correctly restored after reading pass phrase
Migrated from [rt.openssl.org#3935](https://rt.openssl.org/Ticket/Display.html?id=3935) (status was 'new')

Requestors:
* William.Freeman@dealertrack.com

From william.freeman@dealertrack.com on 2015-07-10 20:36:38:
```
I use openssl inside an emacs shell window.  Emacs runs the tty with echo off, collects the line I'm typing (letting me edit it with emacs commands), then sends the whole line when I hit enter.  Since the line as I typed it is already on the screen, I don't need the tty to echo it, or I'll see two copies.

When openssl reads a pass phrase, it turns off echo.  That's a good thing, in general.  (Emacs recognizes the password prompt, and collects the password in a separate window, masking by echoing asterisk for each character, and sends the pass phrase to the tty when collected.)

But then, since, without, apparently, checking, it believes that it turned echo off, openssl unconditionally turns echo on.  This means my subsequent commands (or inputs) appear twice, until I run "stty -echo".

A user of a half duplex terminal (if one can still find any) would be similarly offended.

What is needed is for openssl to record the state of the echo flag before turning it off, and then, after the pass phrase is read, only turn it back on if it was on before.
```

### PR DESCRIPTION
**Pull request by levitte**
The Console UI method always set echo on after prompting without
echo.  However, echo might not have been on originally, so just
restore the original TTY settings.

Fixes #2373

### CODE REVIEW

**Review by richsalz — DISMISSED** [review]

**Comment by dot-asm** [comment]
> @richsalz approved

Really? With all relevant red crosses from CIs?

**Comment by richsalz** [comment]
good point.  should have waited.

**Comment by levitte** [comment]
Fixed

File: include/openssl/uierr.h [hunk 1]
```diff
@@ -20,7 +20,7 @@ int ERR_load_UI_strings(void);
  * UI function codes.
  */
 # define UI_F_CLOSE_CONSOLE                               115
-# define UI_F_ECHO_CONSOLE                                116
+# define UI_F_RESTORE_CONSOLE                             116
```
[the remark this answers is not in the document]
Comment (levitte) [reply include/openssl/uierr.h]: Yeah, good point.  I'll restore the function name then.

File: crypto/err/openssl.txt [hunk 2]
```diff
@@ -1571,7 +1571,7 @@ TS_F_TS_VERIFY:108:*
 TS_F_TS_VERIFY_CERT:109:ts_verify_cert
 TS_F_TS_VERIFY_CTX_NEW:144:TS_VERIFY_CTX_new
 UI_F_CLOSE_CONSOLE:115:close_console
-UI_F_ECHO_CONSOLE:116:echo_console
+UI_F_ECHO_CONSOLE:116:restore_console
```
Comment (dot-asm) [inline crypto/err/openssl.txt]: This needs to be reverted.

File: crypto/ui/ui_openssl.c [hunk 3]
```diff
@@ -190,8 +190,8 @@ static int read_string(UI *ui, UI_STRING *uis);
 static int write_string(UI *ui, UI_STRING *uis);
 
 static int open_console(UI *ui);
-static int echo_console(UI *ui);
 static int noecho_console(UI *ui);
+static int echo_console(UI *ui);
```
Comment (dot-asm) [inline crypto/ui/ui_openssl.c]: While you are at it, maybe move echo_console back, so that there are no unnecessary changes... Note "maybe", as I don't actually insist...

File: crypto/ui/ui_openssl.c [hunk 4]
```diff
@@ -534,7 +533,6 @@ static int echo_console(UI *ui)
 # if defined(_WIN32) && !defined(_WIN32_WCE)
     if (is_a_tty) {
         tty_new = tty_orig;
-        tty_new |= ENABLE_ECHO_INPUT;
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/ui/ui_openssl.c]: I have *no* idea what you're talking about.  This particular line is equally removed in master and in the 1.1.0 branch.  As for #6158, 1.0.2 doesn't have any Win32-specific code here...

[the remark this answers is not in the document]
Comment (dot-asm) [reply crypto/ui/ui_openssl.c]: I think I've mixed up things... Sorry about confusion...

File: crypto/ui/ui_openssl.c [hunk 5]
```diff
@@ -503,7 +503,6 @@ static int echo_console(UI *ui)
 {
 # if defined(TTY_set) && !defined(OPENSSL_SYS_VMS)
     memcpy(&(tty_new), &(tty_orig), sizeof(tty_orig));
-    tty_new.TTY_FLAGS |= ECHO;
 # endif
 
 # if defined(TTY_set) && !defined(OPENSSL_SYS_VMS)
```
Comment (dot-asm) [inline crypto/ui/ui_openssl.c]: Is it a no on removing these two pre-processor directives? I mean both #ifs are equivalent, so it's readability issue, which becomes even more acute with memcpy alone in #if-#endif...

File: crypto/ui/ui_openssl.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/ui/ui_openssl.c]: That's all right

File: crypto/ui/ui_openssl.c (hunk 5 above)
Comment (levitte) [reply crypto/ui/ui_openssl.c]: > Is it a no on removing these two pre-processor directives?

Nah, just me forgetting.  Will fix, thanks for the reminder.

**Review by dot-asm — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

de03cc92d1f3758ad1a525893c93e16b15e6cf45 UI console: Restore tty settings, do not force ECHO after prompt
